.github/workflows/promote_branch.yaml

```
@@ -78,25 +118,18 @@ jobs:
         if: |
           steps.promote.outcome == 'success' && 
           inputs.send-email-report == true
+        id: generate-report
+        continue-on-error: true
         run: |
-          # Determine source and target branches
-          if [ "${{ inputs.promotion-type }}" = "development-to-staging" ]; then
-            FROM_BRANCH="development"
-            TO_BRANCH="staging"
-          else
-            FROM_BRANCH="staging"
-            TO_BRANCH="production"
-          fi
-          
           # Create secret files securely
           echo "${{ secrets.GEMINI_API_KEY }}" > /tmp/gemini_api_key
           echo "${{ secrets.SMTP_USERNAME }}" > /tmp/smtp_username  
           echo "${{ secrets.SMTP_PASSWORD }}" > /tmp/smtp_password
           echo "${{ secrets.GH_TOKEN }}" > /tmp/gh_token
           chmod 600 /tmp/gemini_api_key /tmp/smtp_username /tmp/smtp_password /tmp/gh_token
           
-          # Generate the promotion report
-          python .github/scripts/generate_promotion_report.py $FROM_BRANCH $TO_BRANCH
+          # Generate the promotion report using captured commit range
+          python .github/scripts/generate_promotion_report.py $FROM_BRANCH $TO_BRANCH --commit-range "$COMMIT_RANGE"
```

shouldn't it be COMMITS instead of COMMIT_RANGE?

```suggestion
          python .github/scripts/generate_promotion_report.py $FROM_BRANCH $TO_BRANCH --commit-range "$COMMITS"
```

---

Hi Elena,

No, we should use `COMMIT_RANGE` here. Here's why:

- `COMMIT_RANGE` = `origin/production..origin/staging` (git range format)
- `COMMITS` = multi-line list of individual commit hashes

The script's `get_commits_with_diff()` method expects a git range format (e.g., `origin/production..origin/staging`) to work with the `git log` command. 

The `COMMITS` variable is stored primarily for reference and potential future use in manual recovery scenarios, but the script uses the range format for automated execution.

---

You are almost right, but if I remember correctly, the fetch should work as follows: 
         when you execute https://github.com/konflux-ci/release-service-catalog/blob/development/.github/scripts/generate_promotion_report.py#L119, this fetch should update the references to the post-promotion state, and the range then evaluates to empty and you will get an empty report list.

So if you prefer to use $COMMIT_RANGE, we need to avoid this fetch
.
Or, as I already mentioned, we can use $COMMITS instead.
The Python script supports processing individual commits (lines 141-157). You just need to modify the $COMMITS format (add  | tr '\n' ' '), promote_branch.yaml, line 74:
`
COMMITS=$(git log --pretty=format:"%H" $COMMIT_RANGE | head -50 | tr '\n' ' ')
`
or define some new variable (that will be more correct I think, something like COMMITS_LIST...


You probably test it with the dry-run option (correct me if I'm wrong); such a test doesn't actually test the failure scenario because dry-run prevents the conditions that cause the failure (branches becoming identical). 
The issue will appear on the first real promotion with email reporting.

---

Hi Elena,

Thank you for the thorough review! I've applied your suggested fix (Option 1). Here's what was updated:

**Changes:**
Modified `get_commits_with_diff()` to conditionally skip fetch when `commit_range` is provided.

**How it works:**
```python
if not commit_range:  # Only fetch when range NOT provided
    self.run_git_command(["fetch", "--no-tags", "origin"])
```

**Flow:**
1. Workflow captures `COMMIT_RANGE="origin/production..origin/staging"` BEFORE promotion
2. Promotion happens → remote branches become identical
3. Script receives pre-captured range and **skips fetch**
4. Runner's git refs remain at pre-promotion state (not updated from remote)
5. Range string evaluates using stale refs → shows exactly what was promoted ✓